src/sinks/util/auto_concurrency/tests.rs

```diff
@@ -176,38 +232,45 @@ impl Service<Vec<Event>> for TestSink {
     fn call(&mut self, _request: Vec<Event>) -> Self::Future {
         let now = Instant::now();
         let mut stats = self.stats.lock().expect("Poisoned stats lock");
-        stats.in_flight.adjust(1, now.into());
+        stats.start_request(now);
         let in_flight = stats.in_flight.level();
-
-        let params = self.params;
-        let delay = Duration::from_secs_f64(
-            params.delay
-                * (1.0
-                    + (in_flight - 1) as f64 * params.concurrency_scale
-                    + thread_rng().sample(Exp1) * params.jitter),
-        );
-
-        if params.concurrency_drop > 0 && in_flight >= params.concurrency_drop {
-            stats.in_flight.adjust(-1, now.into());
-            Box::pin(pending())
-        } else {
-            let stats2 = Arc::clone(&self.stats);
-            Box::pin(async move {
-                delay_until(now + delay).await;
-                let mut stats = stats2.lock().expect("Poisoned stats lock");
-                let in_flight = stats.in_flight.level();
-                stats.in_flight.adjust(-1, Instant::now().into());
-
-                if params.concurrency_defer > 0 && in_flight >= params.concurrency_defer {
-                    Err(Error::Deferred)
-                } else {
-                    Ok(Response::Ok)
-                }
-            })
+        let rate = stats.requests.len();
+
+        let action = self
+            .params
+            .concurrency
+            .action_at_level(in_flight)
+            .or_else(|| self.params.rate.action_at_level(rate));
+        match action {
+            None => {
+                let delay = self.delay_at(in_flight, rate);
+                respond_after(Ok(Response::Ok), delay, Arc::clone(&self.stats))
+            }
+            Some(Action::Defer) => {
+                let delay = self.delay_at(1, 1);
+                respond_after(Err(Error::Deferred), delay, Arc::clone(&self.stats))
+            }
+            Some(Action::Drop) => {
+                stats.end_request(now, false);
+                Box::pin(pending())
```

Comment:
Actually, after working in it a bit, we want to be able to model both behaviors. Yes, this pending request will eventually be dropped, not because it falls out of scope but because of the `Timeout` layer that is wrapped around the test sink. Modelling this timeout behavior is different than an immediate closed connection, which can be modelled by a fast defer.

Reply:
Ah, gotcha, yeah that behavior could be useful to model too. Maybe we could call it `Action::Timeout` instead?

Reply:
I'm not sure `timeout` is accurate, as it describes the effect instead of the direct action, but I see where you're going.